### ISSUE TITLE
EpiCSeg segment: error invalid length

### ISSUE THREAD

**Issue by mpi-pebert** [lamortenera/epicseg#18]
Hi,
I am trying to run EpiCSeg on a somewhat large dataset; counting and normalization run fine, but the `segment` command fails stating the following:
```R
Creating a generic function for 'nchar' from package 'base' in package 'S4Vectors'
    concatenating count matrices
    Error: invalid length
    Execution halted
```
My dataset consists of 38 samples (count matrices), summing up to a total of 588.113.802 200bp bins - is that, for some reason, too much data? Or is the error above unrelated to the amount of data?

+Peter

**Comment by lamortenera** [comment]
Hi Peter,

this is almost surely a data overflow problem. The overflow probably
happens already in the EpiCSeg code at line 193 of

epicseg
<https://github.com/lamortenera/epicseg/tree/3816b45a2231af9438fbb1b777f80659deb76982>
/src
<https://github.com/lamortenera/epicseg/tree/3816b45a2231af9438fbb1b777f80659deb76982/src>
/misc.cpp

But even if I would fix it (i.e. replace nbins*ncounts with something like
((long) nbins)*ncounts) I don't think it is possible to allocate such a
matrix (Rcpp doesn't have a matrix constructor that takes longs). In R the
dimensions of a matrix (or of a multidimensional array) are stored as a
32bit int, so no dimension can be larger than 2^31 - 1, even though array
based on long vectors can have a number of elements (the product of all
dimensions) of 2^52. This is what is written here:

https://stat.ethz.ch/R-manual/R-devel/library/base/html/LongVectors.html

Basically if nsamples and nbins are the number of samples and the number of
bins, respectively, you can't have nsamples * nbins >= n^31.
By curiosity, where did you find a genome with more than 200 * 588.113.802
base pairs? You should be able to concatenate something like 200 human
genomes, independently on the number of histone marks (provided they are
less than 2^31 :D) , but each human genome should have less than 15.000.000
bins.

Sorry for that, I can imagine that it is very annoying.

Cheers
Ale

**Comment by mpi-pebert** [comment]
Hi Ale,
sorry, probably, my error description was not precise enough - the dataset consists of
38 human samples, each with
6+1 histone marks, resulting in
15.476.679 genomic bins per sample, giving a total of
588.113.802 bins for the entire dataset
Now, if I am not mistaken, `588.113.802 << 2^31 - 1`, which would imply that it *should* work, as you wrote:

*You should be able to concatenate something like 200 human genomes, independently on the number of histone marks*

So, if *each* dimension can be as large as `2^31 - 1`, then the error is coming from somewhere else - that would be good news for me and bad news for you :-(
I would appreciate you giving this another thought if time permits.

Cheers
Peter

**Comment by lamortenera** [comment]
Hi Peter,
thanks for insisting, I misread your original message, and I think I also
found the bug. It must be line 132 of misc.cpp, where I am summing up
integers into an accumulator that will overflow. It will be easy to write a
fix that should work, it will be harder to test it. I don't have access to
the MPI cluster anymore, nor to the Illumina one, so I'll need to get
creative :D Probably I can test it with travis.

The fix could be ready even tomorrow or the day after tomorrow, for the
tests you'll have to wait till Monday.

Cheers
Ale

**Comment by mpi-pebert** [comment]
Great, please let me know when it's fixed!
Thanks for your help.

+Peter

**Comment by mpi-pebert** [comment]
Any updates?

**Comment by lamortenera** [comment]
Sorry I still didn't get around that. I'll try to work on it this weekend.

**Comment by lamortenera** [comment]
Hi Peter,

I pushed a fix, but I am unable to test it properly, I don't have access to
machines with so much memory. Could you try to see if it works?

Cheers
Alessandro

**Comment by ghost** [comment]
Hi Ale,
thanks, I will give it a try as soon as possible; apparently, some library versions of our R installation have changed and now I am in conflict resolution mode... do you happen to have a listing with all package dependencies/versions known to work? Maybe I just set up a clean, new conda environment, might be faster...
Best,
Peter

**Comment by lamortenera** [comment]
Hi Peter,

unfortunately I don't have epicseg installed in any machine that I can
control, but Travis already prints some info about the used packages:

https://travis-ci.org/lamortenera/epicseg/jobs/254113430
click on: R-installed-versions

And you'll see a list of all R packages with their versions.

But I think that if you start with a clean conda with R 3.2 or 3.3 and
follow the installation instructions for epicseg everything should work.

I hope that helps

**Comment by ghost** [comment]
Ok, this turns out to become quite an annoying thing... I have a fresh environment, essentials as follows
```yaml
r-base=3.3.2
bioconductor-biocinstaller=1.24.0
r-devtools=1.12.0
```
Overall, installation worked fine, but for the github packages, I had to set the following option:
```R
# if this is empty, need to set the option below
> getOption("unzip")
[1] ""

options(unzip = "unzip")
```
Found this here: https://github.com/hadley/devtools/issues/406

Installing the `kfoots` package then works fine, yet when I try to install `epicseg`, I get the following:

```R
> install_github("lamortenera/epicseg")
Downloading GitHub repo lamortenera/epicseg@master
from URL https://api.github.com/repos/lamortenera/epicseg/zipball/master
Installing epicseg
trying URL 'https://cran.rstudio.com/src/contrib/RColorBrewer_1.1-2.tar.gz'
Content type 'application/x-gzip' length 11532 bytes (11 KB)
==================================================
downloaded 11 KB

sh: 1: /bin/gtar: not found
sh: 1: /bin/gtar: not found
Error in system(cmd, intern = TRUE) : error in running command
In addition: Warning message:
In utils::untar(src, exdir = target, compressed = "gzip") :
  '/bin/gtar -xf '/tmp/RtmpBVbJ2D/RColorBrewer_1.1-2.tar.gz' -C '/tmp/RtmpBVbJ2D/devtools9b2f20854ce3'' returned error code 127
```
Ok, `gtar` does not exist on our servers and I cannot simply symlink to `/bin/tar` because I am not root; is that a problem with `devtools` or with `epicseg`? Any idea for a workaround?

**Comment by lamortenera** [comment]
Hi Peter,

I am sorry for all this pain.. I never had to do anything like that.
It seems like you have some non-standard settings (never seen a linux
environment without unzip and untar). I would try 2 things:
1. I don't know if you tried that already, but "conda install
r-essentials" should install a bunch of common packages, hopefully
including things like unzip and untar. Could you try to start from
there?
2. You could see how utils::untar works by typing:
> utils::untar
It uses an environment variable called $TAR. So if in a bash shell you do:
> export TAR=/path/to/my/tar
and then launch R, the utils::untar command should work.

I hope that helps, keep me updated.
Cheers
Alessandro

**Comment by ghost** [comment]
Hi Ale,
thanks for the help.
Well, our servers are sporting a down-to-earth Debian 7.11, I guess it does hardly get any more basic than that. But please note that in the above message, I was talking about `gtar` with a `g` in front - to be honest, I have never seen that before...

Back to topic...
1) r-essentials: does not change anything
2) I exported `TAR=/bin/tar` in my environment, no change

I read again about this weird `unzip` thing and noted that people that provided help report the value `internal` for their R session; just the *first* workaround in the issue linked above is using `unzip`. So the following leads to a successful installation:
```R
options(unzip = "internal")
```
Still not sure where this problem is coming from in the first place (devtools?), but now the setup is complete.
So, setup complete, I'll run the analysis again tomorrow and let you know the result. Thanks again.
Cheers,
Peter

**Comment by lamortenera** [comment]
I am going to call you "the unstoppable" :)

Cheers
Ale

**Comment by ghost** [comment]
It seems that your fix worked and the `segment` command proceeds now further; now this happens:
```R
concatenating count matrices
    Error in kfoots_error_handler(e$message, regions, binsize) : 
      cannot allocate vector of size 30.7 Gb
    Calls: <Anonymous> ... tryCatchList -> tryCatchOne -> <Anonymous> -> kfoots_error_handler
    Execution halted
    
    The stdout was: 
    loading epicseg
    [init] performing PCA on the count matrix 
```
I ran this two times and watched the memory consumption on the server during the second run - the process occupied around 60GB, leaving another ~60GB free and no other processes were running. So, by what I can tell, it should have been possible to at least allocate another 30GB. Any idea?
Cheers,
Peter

**Comment by lamortenera** [comment]
Hi Peter, sorry for the late reply.
I think the error message doesnt leave much doubt, R wasnt able to get the
memory it needed. If you monitor the ram usage with top that's not always
reliable. I think the usr/bin/time function has a -m or -s flag that
reports the peak memory usage... then who knows if the os or the server
configuration lets epicseg take as much memory as needed or if it reserves
some memory for other stuff .

You could do the following (assuming you don't have a server with more
memory):
1. run epicseg only on chromosome 1 for all genomes (that should be about a
tenth of the data)
2. use the trained model and the --notrain option to calculate the
segmentation on the whole genome of each sample separately.

There is a small annoyance because you need to call normalizecounts. The
laziest thing to do is to count the reads from the BAM files and call
normalize counts twice, first for chromosome 1 and for training, and then
for the whole genome and for calculating the states. But then the counts
may be normalized slightly differently so you don't train exactly on the
same data where you calculate the segmentation.
The most appropriate thing to do is to count and normalize once on the
whole genome, and then with an awk/sed/python/bash script select only those
bins from the count matrices that are related to chromosome 1, and write
those counts to other smaller count matrices that will be used for training.
It shouldn't be difficult: If in your regions bed file the first line
represents chromosome 1 and has size 200*n, then the first n lines of each
count matrix file are the ones you need for training.

Sorry for the difficulties

Keep me updated
Alessandro Mammana




On 20 Jul 2017 17:34, "Peter Ebert" <notifications@github.com> wrote:

It seems that your fix worked and the segment command proceeds now further;
now this happens:

concatenating count matrices
    Error in kfoots_error_handler(e$message, regions, binsize) :
      cannot allocate vector of size 30.7 Gb
    Calls: <Anonymous> ... tryCatchList -> tryCatchOne -> <Anonymous>
-> kfoots_error_handler
    Execution halted

    The stdout was:
    loading epicseg
    [init] performing PCA on the count matrix

I ran this two times and watched the memory consumption on the server
during the second run - the process occupied around 60GB, leaving another
~60GB free and no other processes were running. So, by what I can tell, it
should have been possible to at least allocate another 30GB. Any idea?
Cheers,
Peter

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/lamortenera/epicseg/issues/18#issuecomment-316741605>,
or mute the thread

**Comment by ghost** [comment]
Hi Ale,
thanks for your input. Based on past experience, I would not assume that the server config or OS is an issue; R processes have overloaded servers on numerous occasions... but anyway, I'll take your advice and train EpiCSeg on a subset of the data, I guess this is the most straightforward solution to the problem.
Cheers,
Peter

**Comment by ghost** [comment]
Status update: I trained a model on the first two chromosomes in the regions file (chr1 and chr10), and supplied the model and the `--notrain` parameter in a subsequent segmentation step - this time, `kfoots` tells me that it fails to allocate a vector of size ~39GB. I will give it another try with just chr1 and let you know how it works out...

**Comment by ghost** [comment]
Hm, so with just chr1, it seems that EpiCSeg is proceeding (compared to the whole genome dataset) - I can now observe the memory consumption peaking at around 100GB and then it crashes....
```R
concatenating count matrices
    Error in kfoots_error_handler(e$message, regions, binsize) : 
      cannot allocate vector of size 39.4 Gb
    Calls: <Anonymous> ... tryCatchList -> tryCatchOne -> <Anonymous> -> kfoots_error_handler
    In addition: Warning message:
    In (function (counts, regions, nstates = NULL, model = NULL, notrain = FALSE,  :
      collapsing the initial probabilities for compatibility with the new set of observations
    Execution halted
```
This now is clear, our servers are "only" equipped with 128GB, no way that this could fly.
Is there any chance that R does unnecessary copies of objects, i.e., would it affect the situation to just run this on a single core? If not, I guess I have to reduce the dataset... :-(

**Comment by ghost** [comment]
Ok, I answered that last part myself - apparently not :-)
